presto-common/src/main/java/com/facebook/presto/common/block/Int128ArrayBlockBuilder.java

```diff
@@ -0,0 +1,459 @@
+     *                       For example, offset = 1 means the next position (one 128-bit integer or 16 bytes) to the specified position.
+     *                       This means we always compare bytes starting at 128-bit integer boundaries.
+     * @param length The length in bytes. It has to be a multiple of 16.
+     */
+    @Override
+    public Slice getSliceUnchecked(int internalPosition, int offset, int length)
+    {
+        int num128Integers = getNum128Integers(length);
+        return Slices.wrappedLongArray(values, (internalPosition + offset) * 2, num128Integers * 2);
+    }
+
+    @Override
+    public int getSliceLengthUnchecked(int position)
+    {
+        return SIZE_OF_LONG * 2;
+    }
+
+    /**
+     * Is the byte sequences at the {@code position + offset} position in the 128-bit values of {@code length} bytes equal
+     * to the byte sequence at {@code otherOffset} in {@code otherSlice}.
+     *
+     * @param position The position of 128-bit integer.
+     * @param offset The offset to the position in the unit of 128-bit integers.
+     * For example, offset = 1 means the next position (one 128-bit integer or 16 bytes) to the specified position.
+     * This means we always compare starting at 128-bit integer boundaries.
+     * @param otherSlice The slice to compare to.
+     * @param otherOffset The offset in bytes to the start of otherSlice.
+     * @param length The length to compare in bytes. It has to be a multiple of 16.
+     * @return True if the bytes are the same, false otherwise.
+     */
+    @Override
+    public boolean bytesEqual(int position, int offset, Slice otherSlice, int otherOffset, int length)
+    {
+        int num128Integers = getNum128Integers(length);
+        checkValidRegion(positionCount, position + offset, num128Integers);
+        if (otherOffset < 0 || length < 0 || otherOffset + length > otherSlice.length()) {
+            throw new IllegalArgumentException(format("otherOffset %d, length %d are invalid for otherSlice with length %d", otherOffset, length, otherSlice.length()));
+        }
+
+        int currentPosition = (position + offset + getOffsetBase()) * 2;
+        for (int i = 0; i < num128Integers; i++) {
+            if (values[currentPosition] != otherSlice.getLong(otherOffset) || values[currentPosition + 1] != otherSlice.getLong(otherOffset + SIZE_OF_LONG)) {
+                return false;
+            }
+
+            currentPosition += 2;
+            otherOffset += SIZE_OF_LONG * 2;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean mayHaveNull()
+    {
+        return hasNullValue;
+    }
+
+    @Override
+    public boolean isNull(int position)
+    {
+        checkReadablePosition(position);
+        return valueIsNull[position];
+    }
+
+    @Override
+    public void writePositionTo(int position, BlockBuilder blockBuilder)
+    {
+        checkReadablePosition(position);
+        blockBuilder.writeLong(values[position * 2]);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).